Source/FieldSolver/FiniteDifferenceSolver/HybridPICModel/HybridPICModel.cpp

```diff
@@ -534,3 +534,156 @@ void HybridPICModel::FillElectronPressureMF (
+            Bfield[lev][ii]->nGrowVect()
+        );
+        K[ii].setVal(0.0);
+    }
+
+    // The Runge-Kutta scheme begins here.
+    // Step 1:
+    FieldPush(
+        Bfield, Efield, Jfield, rhofield, edge_lengths,
+        0.5_rt*dt, dt_type, ng, nodal_sync
+    );
+
+    // The Bfield is now given by:
+    // B_new = B_old + 0.5 * dt * [-curl x E(B_old)] = B_old + 0.5 * dt * K0.
+    for (int ii = 0; ii < 3; ii++)
+    {
+        // Extract 0.5 * dt * K0 for each direction into index 0 of K.
+        MultiFab::LinComb(
+            K[ii], 1._rt, *Bfield[lev][ii], 0, -1._rt, B_old[ii], 0, 0, 1, ng
+        );
+    }
+
+    // Step 2:
+    FieldPush(
+        Bfield, Efield, Jfield, rhofield, edge_lengths,
+        0.5_rt*dt, dt_type, ng, nodal_sync
+    );
+
+    // The Bfield is now given by:
+    // B_new = B_old + 0.5 * dt * K0 + 0.5 * dt * [-curl x E(B_old + 0.5 * dt * K1)]
+    //       = B_old + 0.5 * dt * K0 + 0.5 * dt * K1
+    for (int ii = 0; ii < 3; ii++)
+    {
+        // Subtract 0.5 * dt * K0 from the Bfield for each direction, to get
+        // B_new = B_old + 0.5 * dt * K1.
+        MultiFab::Subtract(*Bfield[lev][ii], K[ii], 0, 0, 1, ng);
+        // Extract 0.5 * dt * K1 for each direction into index 1 of K.
+        MultiFab::LinComb(
+            K[ii], 1._rt, *Bfield[lev][ii], 0, -1._rt, B_old[ii], 0, 1, 1, ng
+        );
+    }
+
+    // Step 3:
+    FieldPush(
+        Bfield, Efield, Jfield, rhofield, edge_lengths,
+        dt, dt_type, ng, nodal_sync
+    );
+
+    // The Bfield is now given by:
+    // B_new = B_old + 0.5 * dt * K1 + dt * [-curl  x E(B_old + 0.5 * dt * K1)]
+    //       = B_old + 0.5 * dt * K1 + dt * K2
+    for (int ii = 0; ii < 3; ii++)
+    {
+        // Subtract 0.5 * dt * K1 from the Bfield for each direction to get
+        // B_new = B_old + dt * K2.
+        MultiFab::Subtract(*Bfield[lev][ii], K[ii], 1, 0, 1, ng);
+    }
+
+    // Step 4:
+    FieldPush(
+        Bfield, Efield, Jfield, rhofield, edge_lengths,
+        0.5_rt*dt, dt_type, ng, nodal_sync
+    );
+
+    // The Bfield is now given by:
+    // B_new = B_old + dt * K2 + 0.5 * dt * [-curl x E(B_old + dt * K2)]
+    //       = B_old + dt * K2 + 0.5 * dt * K3
+    for (int ii = 0; ii < 3; ii++)
+    {
+        // Subtract B_old from the Bfield for each direction, to get
+        // B = dt * K2 + 0.5 * dt * K3.
+        MultiFab::Subtract(*Bfield[lev][ii], B_old[ii], 0, 0, 1, ng);
+
+        // Add dt * K2 + 0.5 * dt * K3 to index 0 of K (= 0.5 * dt * K0).
+        MultiFab::Add(K[ii], *Bfield[lev][ii], 0, 0, 1, ng);
+
+        // Add 2 * 0.5 * dt * K1 to index 0 of K.
+        MultiFab::LinComb(
+            K[ii], 1.0, K[ii], 0, 2.0, K[ii], 1, 0, 1, ng
+        );
+
+        // Overwrite the Bfield with the Runge-Kutta sum:
+        // B_new = B_old + 1/3 * dt * (0.5 * K0 + K1 + K2 + 0.5 * K3).
+        MultiFab::LinComb(
+            *Bfield[lev][ii], 1.0, B_old[ii], 0, 1.0/3.0, K[ii], 0, 0, 1, ng
+        );
```

Comment:
Would it be worth to write all these operations in a single `ParallelFor` kernel, both for readability and to avoid kernel-launch overheads from multiple kernels?

Reply:
Thanks for the suggestion. I also thought about that as a good follow-up performance improvement but haven't gotten around to it yet since with this new scheme the field solver takes a small amount of runtime in the simulations we are commonly running (so it is not high on the priority list). But of course it would be a good thing to make the code more performant. My current idea is to save this for a time when we have a new team member for which this would be a good task as a way to learn the hybrid-PIC algorithm as well as how to work with WarpX/AMReX.